### PR TITLE
php8

### PR DESCRIPTION
**Pull request by mattmary**
https://innocraft.atlassian.net/browse/DEV-2573

### CODE REVIEW

**Review by tsteur — COMMENTED** [review]
@mattmary there's an `echo` in there and some line commented. This might need to be adjusted? Did the UI work nicely with PHP 8 without changes?

**Comment by mattmary** [comment]
@tsteur I don't understand how to resolve these conflicts. it's not clear in github. Can you help me?

**Comment by justinvelluppillai** [comment]
I can help @mattmary

**Review by tsteur — COMMENTED** [review]
@mattmary the generated docs look quite different now. It seems it fails to parse a few things like the namespaces and seeing lots of classes replaced with `Stmt_Namespace`?

**Comment by tsteur** [comment]
Any update on this one @mattmary ?

**Comment by mattmary** [comment]
no I didn't find a solution for now: the parsing library has been updated
but there are a lot of change in it. I have to understand how the parsing
is done and how the new library works.
it will take some time I guess.

**Comment by mattmary** [comment]
hi @tsteur I'm blocked for now by https://github.com/matomo-org/matomo/issues/19553

**Comment by tsteur** [comment]
@mattmary it seems this has been merged now. Is it maybe working again?

**Comment by mattmary** [comment]
@tsteur yes., I can generate the documentation now.

**Comment by mattmary** [comment]
Hi @tsteur 

I think I found the explanation: In the NodeVisitor class the Namespace is not well set: when using in line 47 
`$this->context->enterNamespace((string) $node->name);`
instead of 
`$this->context->enterNamespace((string) $node->getType());`
I have the appropriate Namespace specified when I browse http://local.developer.matomo.org/api-reference/. 
Is it where we can list the namespaces? I've browsed my local documentation and it's the only location I found where the namespace is displayed so far.

So I wanted to push the update on tsteur/Sami repository on the branch 4.x-devphp8. Unfortunately, I don't see a dev-4.x-devphp8 on this repository. Is it a clone? Do I miss something? Does the branches in the main repository are always available in the child repositories?

I've seen that there is a matomo/Sami repository with a branch 4.x-devphp8 which includes this update. Is it the case, and if the cloned repositories get the main repositories branches, why I don't see this update in my local environment?

So I've tried to update the generator/composer.json to replace the component `tsteur/Sami` by `matomo/Sami`,  like this:

`{
    "name": "piwik/pluginapi-doc-generator",
    "description": "Generates plugin API documentation",
    "keywords": ["matomo","web","analytics"],
    "homepage": "https://matomo.org",
    "license": "GPL-3.0+",
    "authors": [
        {
            "name": "The Matomo Team",
            "email": "hello@matomo.org",
            "homepage": "https://matomo.org/the-piwik-team/"
        }
    ],
    "autoload": {
        "psr-4": {
            "Piwik\\Plugins\\": "../piwik/plugins/",
            "Piwik\\": "../piwik/core/"
        },
        "psr-0": {
            "Zend_": "../piwik/libs/",
            "HTML_": "../piwik/libs/",
            "PEAR_": "../piwik/libs/",
            "Archive_": "../piwik/libs/"
        }
    },
    "repositories": [
        {
            "type": "git",
            "url": "https://github.com/matomo/Sami"
        }
    ],
    "require": {
        "php": ">=5.3.0",
        "matomo/sami": "dev-4.x-devphp8"
    }
}
`
But it asks me a credential when I run `composer update`.

Am I doing something wrong? What is missing? 

@justinvelluppillai do you have an idea too?

**Comment by tsteur** [comment]
> Is it where we can list the namespaces? I've browsed my local documentation and it's the only location I found where the namespace is displayed so far.

We're only showing the [list of classes](https://developer.matomo.org/api-reference/classes) but not the list of namespaces as far as I know 👍 

> So I wanted to push the update on tsteur/Sami repository on the branch 4.x-devphp8. Unfortunately, I don't see a dev-4.x-devphp8 on this repository. Is it a clone? Do I miss something? Does the branches in the main repository are always available in the child repositories?
> I've seen that there is a matomo/Sami repository with a branch 4.x-devphp8 which includes this update. Is it the case, and if the cloned repositories get the main repositories branches, why I don't see this update in my local environment?
> So I've tried to update the generator/composer.json to replace the component tsteur/Sami by matomo/Sami, like this: ...
> Am I doing something wrong? What is missing?

Generally, it should already use https://github.com/matomo-org/Sami because of this line https://github.com/matomo-org/developer-documentation/blob/live/generator/composer.json#L29 The package name doesn't need to be changed AFAIK

I'm not sure though why it does need credentials. It works for me without issues when using the composer you already have in the branch (not any changed composer) . Does it work when you don't change the composer file? Maybe the change is not needed?
image

**Comment by mattmary** [comment]
Hi @tsteur 

I've cleared the composer cache, rm the vendor folder and now I get the appropriate version of the github repository.
It should work but now I'm facing a new exception:

`./generate.sh 
generating 4.x-dev - 4.x
Updated 233 paths from the index
Already on '4.x-dev'
Your branch is up to date with 'origin/4.x-dev'.
4.x-dev
From github.com:matomo-org/piwik
 * branch                  4.x-dev    -> FETCH_HEAD
Already up to date.
You are already using the latest available Composer version 2.3.10 (stable channel).
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Package lox/xhprof is abandoned, you should avoid using it. No replacement was suggested.
Package phpunit/php-token-stream is abandoned, you should avoid using it. No replacement was suggested.
Generating autoload files
32 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
There was an error during the initialisation but we try to proceed anyway. FYI: error was 
The configuration file {/var/www/matomo/matomo/developer-documentation/piwik/config/config.ini.php} has not been found or could not be read.
 » Please check that /var/www/matomo/matomo/developer-documentation/piwik/config/config.ini.php is readable by the user 'matthieu'. 
 

PHP Fatal error:  Uncaught TypeError: Illegal offset type in /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Reflection/ClassReflection.php:405
Stack trace:
#0 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(242): Sami\Reflection\ClassReflection->addMethod()
#1 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(61): Sami\Parser\NodeVisitor->addMethod()
#2 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(200): Sami\Parser\NodeVisitor->enterNode()
#3 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(114): PhpParser\NodeTraverser->traverseArray()
#4 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(223): PhpParser\NodeTraverser->traverseNode()
#5 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(114): PhpParser\NodeTraverser->traverseArray()
#6 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(223): PhpParser\NodeTraverser->traverseNode()
#7 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(91): PhpParser\NodeTraverser->traverseArray()
#8 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/CodeParser.php(43): PhpParser\NodeTraverser->traverse()
#9 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/Parser.php(53): Sami\Parser\CodeParser->parse()
#10 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(466): Sami\Parser\Parser->parse()
#11 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(109): Sami\Project->parseVersion()
#12 /var/www/matomo/matomo/developer-documentation/generator/generate.php(111): Sami\Project->update()
#13 /var/www/matomo/matomo/developer-documentation/generator/generate.php(144): generateApiClassesReference()
#14 {main}
  thrown in /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Reflection/ClassReflection.php on line 405`

this error is fixed if I replace in ClassReflection the line 
`    public function addMethod(MethodReflection $method)
    {
        $this->methods[$method->getName()] = $method;
        $method->setClass($this);
    }
`
by
`    public function addMethod(MethodReflection $method)
    {
        $this->methods[(string) $method->getName()] = $method;
        $method->setClass($this);
    }
`
when I fix it, I have others errors on undefined properties.

does this package tsteur/sami comes from an external source or this is something you developed? 

can we update it?

**Comment by tsteur** [comment]
> does this package tsteur/sami comes from an external source or this is something you developed?
> can we update it?


@mattmary the code for this should be in https://github.com/matomo-org/Sami/tree/4.x-devphp8 it is a fork of https://github.com/FriendsOfPHP/Sami . We can change the code in there and update it

**Comment by mattmary** [comment]
Hello @tsteur 

I'm not sure if I follow the right way to fix this issue. Can you give me your feedback?

To summarise, to update the php8 I needed to use your new component "tsteur/sami": "dev-4.x-devphp8",
It loads the new parsing lib, compatible with php8.
Unfortunately, this parsing lib has been upgraded to a major version and there is no backward compatibility: objects definitions are not the same one. This API uses arrays to stored parsed data and now, as it returns objects, the indexes need to be cast as string to be stored in arrays. I've made a lot of updates in your tsteur/sami component to fix these issues.
But now after fixing all these issues I have errors in the twig component and as far as I've seen there is no workaround to fix it.

I've spent a lot of time on this issue, and I could spend a lot more to fix this twig issue. 

But regarding the time I spent on it and the complexity of this task, I'm not sure if I follow the right way.
Your feedback would be welcome.

**Comment by tsteur** [comment]
> To summarise, to update the php8 I needed to use your new component "tsteur/sami": "dev-4.x-devphp8",

👍  I believe so. I'm not 100% sure though. Note that "tsteur/sami" actually uses this repository: https://github.com/matomo-org/Sami/tree/4.x-devphp8 
That is because of this line: https://github.com/matomo-org/developer-documentation/blob/live/generator/composer.json#L29

Is the latest code committed? What be the quickest way for me to reproduce the problem?

**Comment by mattmary** [comment]
Hello @tsteur 

If you want to reproduce it, I would have to share with you the updates I made in tsteur/sami component.
As this is a composer component, not sure which is the best approach to do so? Create me a new branch based upon the dev-4.x-devphp8 branch and let me commit my update in it?
Is it a solution acceptable for you? Do you see any other approach?

**Comment by tsteur** [comment]
@mattmary can you push the changes into a new branch on https://github.com/matomo-org/Sami ?

**Comment by mattmary** [comment]
Hello @tsteur 
here's the PR https://github.com/matomo-org/Sami/pull/2

**Comment by tsteur** [comment]
@mattmary when I execute it, I get an error like 

```
PHP Fatal error:  Uncaught TypeError: Illegal offset type in isset or empty in generator/vendor/twig/twig/src/Extension/CoreExtension.php:1406
```

Is this what you're getting as well?

**Comment by mattmary** [comment]
Hi @tsteur 

Yes, I get this error too and this is why I requested a different point of view: I've tried to upgrade the twig version without success. So maybe my fixes in the branch are not the appropriate way to solve the PHP8 issues.
These issues come from objects which are stored in arrays. A nice fix would be to replace the array storage into a Collection management, but it would mean rewrite all the library.
This is why I request from you a different point of view.

**Comment by tsteur** [comment]
@mattmary I've debugged and we're now getting objects instead of names when accessing `.name` meaning when we're using eg the `.name` around objects in twig then we first have to convert it to a string.

I've made some changes so that when we access a key on an array/object that we make sure to convert it to an array. This seems to work now for me. Can you check? I haven't checked in details but at least it runs through now

**Comment by tsteur** [comment]
fyi these are the fixes we had applied to Sami originally https://github.com/FriendsOfPHP/Sami/compare/master...matomo-org:Sami:master I assume that similar fixes we may need to apply again to generate correct API output

**Comment by mattmary** [comment]
@tsteur thank you for your feedback. 
I did not make any updates so far in this branch, just waiting for the PR https://github.com/matomo-org/Sami/pull/3 which includes your updates to be merged

File: docs/4.x/generated/Piwik/API/Request.md [hunk 1]
```diff
@@ -120,11 +120,11 @@ object.
 #### Signature
 
 
-- *Returns:*  [`DataTable`](../../Piwik/DataTable.md)|`Piwik\API\Map`|`string` &mdash;
+- *Returns:*  `Piwik\API\DataTable`|`Piwik\API\Map`|`string` &mdash;
```
Comment (tsteur) [inline docs/4.x/generated/Piwik/API/Request.md]: fyi @mattmary these class names are not generated correctly. It should be `Piwik\DataTable` and it should link to the correct page

File: docs/4.x/generated/Piwik/Archive.md [hunk 2]
```diff
@@ -176,7 +176,7 @@ use [build()](/api-reference/Piwik/Archive#build).
        Whether `'all'` sites are being queried or not. If true, then the result of querying functions will be indexed by site, regardless of whether `count($idSites) == 1`.
     - `$isMultipleDate` (`bool`) &mdash;
        Whether multiple dates are being queried or not. If true, then the result of querying functions will be indexed by period, regardless of whether `count($periods) == 1`.
-- It returns a `Piwik\Archive\ArchiveQuery` value.
+- It returns a `Piwik\ArchiveQuery` value.
```
Comment (tsteur) [inline docs/4.x/generated/Piwik/Archive.md]: here the class name is now missing the `\Archive` part

File: docs/4.x/generated/Piwik/ArchiveProcessor.md [hunk 3]
```diff
@@ -78,7 +78,7 @@ data for.
 
 #### Signature
 
-- It returns a [`Parameters`](../Piwik/ArchiveProcessor/Parameters.md) value.
+- It returns a `Piwik\Parameters` value.
```
Comment (tsteur) [inline docs/4.x/generated/Piwik/ArchiveProcessor.md]: Here it's missing the `/ArchiveProcessor` part and a link

File: docs/4.x/generated/Piwik/Columns/Dimension.md [hunk 4]
```diff
@@ -535,7 +535,7 @@ Returns the name of the plugin that contains this Dimension.
 
 - It returns a `string` value.
 - It throws one of the following exceptions:
-    - [`Exception`](http://php.net/class.Exception) &mdash; if the Dimension is not located within a Plugin module.
+    - `Piwik\Columns\Exception` &mdash; if the Dimension is not located within a Plugin module.
```
Comment (tsteur) [inline docs/4.x/generated/Piwik/Columns/Dimension.md]: This one actually throws `\Exception` and not `Piwik\Exception` and neither `Piwik\Columns\Exception` and the link went missing as well. Same issue seems to be happening many times.

File: docs/4.x/generated/Piwik/Plugin/Controller.md [hunk 5]
```diff
@@ -152,7 +152,7 @@ If the date is absolute, ie. YYYY-MM-DD, it will not be converted to the timezon
        `'today'`, `'yesterday'`, `'YYYY-MM-DD'`
     - `$timezone` (`string`) &mdash;
        The timezone to use.
-- It returns a [`Date`](../../Piwik/Date.md) value.
+- It returns a `Piwik\Plugin\Date` value.
```
Comment (tsteur) [inline docs/4.x/generated/Piwik/Plugin/Controller.md]: Here it also should be still `Piwik\Date` as `Piwik\Plugin\Date` doesn't exist and it should link to the page. Same above for `Piwik\Plugin\Site` which should be `Piwik\Site`

**Review by tsteur — COMMENTED** [review]
fyi @mattmary left a few comments on the generated docs. Didn't look into the code. Just noticed that I believe pretty much all classnames seem to have issues that either the name isn't correct and/or the link is missing which was there previously.

File: docs/4.x/generated/Piwik/API/Request.md [hunk 6]
```diff
@@ -5,7 +5,7 @@ Request
 
 Dispatches API requests to the appropriate API method.
 
-The Request class is used throughout Piwik to call API methods. The difference
+okokokThe Request class is used throughout Piwik to call API methods. The difference
```
Comment (lance-matomo) [inline docs/4.x/generated/Piwik/API/Request.md]: I'm not sure why all these changes say ok in them?

Comment (mattmary) [reply docs/4.x/generated/Piwik/API/Request.md]: Sorry, it looks like when I generated the documentation, it did not load the latest vendor version of one component. So my debug traces were still there, and I didn't see them when pushing. You have a new documentation version now.

**Comment by tsteur** [comment]
@mattmary looking at the `hooks.md` there are quite a few changes. One thing I noticed is that hook references to `CustomAlerts` etc are missing. This indicates that the submodules are not initialised.

I would assume that by running `cd piwik && git submodule init` and then generating the docs again via the `generate.sh` script that these references will then show up.

To prevent needing to run this `git submodule init` command when installing the docs locally we could change this `git submodule update` command to also add additionally the option `--init` meaning it be `git submodule update --recursive --force`. This way, if Matomo adds submodules in the future (which it maybe already has), they will be updated automatically too.

Then the custom alerts hook references should be added when generating the docs

**Comment by mattmary** [comment]
Hello @tsteur 

I definitively did not manage manually the submodules, yes.
Following your comment, I've added the --init option to the line `git submodule update --recursive --force` in the generate.sh script. It is then check out the submodules.

But unfortunately, I have an error now when generating the documentation:

`PHP Fatal error:  Uncaught Error: Class "Piwik\Plugins\ActivityLog\Activity\Activity" not found in /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/BaseActivity.php:16
Stack trace:
#0 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include()
#1 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#2 /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/VersionBaseActivity.php(12): Composer\Autoload\ClassLoader->loadClass()
#3 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include('...')
#4 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#5 /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/VersionDeleted.php(11): Composer\Autoload\ClassLoader->loadClass()
#6 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include('...')
#7 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#8 [internal function]: Composer\Autoload\ClassLoader->loadClass()
#9 /var/www/matomo/matomo/developer-documentation/generator/ApiClasses/Filter.php(23): class_exists()
#10 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(140): ApiClasses\Filter->acceptClass()
#11 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(100): Sami\Parser\NodeVisitor->addClassOrInterface()
#12 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(53): Sami\Parser\NodeVisitor->addClass()
#13 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(202): Sami\Parser\NodeVisitor->enterNode()
#14 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(116): PhpParser\NodeTraverser->traverseArray()
#15 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(225): PhpParser\NodeTraverser->traverseNode()
#16 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(93): PhpParser\NodeTraverser->traverseArray()
#17 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/CodeParser.php(43): PhpParser\NodeTraverser->traverse()
#18 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/Parser.php(53): Sami\Parser\CodeParser->parse()
#19 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(466): Sami\Parser\Parser->parse()
#20 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(109): Sami\Project->parseVersion()
#21 /var/www/matomo/matomo/developer-documentation/generator/generate.php(119): Sami\Project->update()
#22 /var/www/matomo/matomo/developer-documentation/generator/generate.php(152): generateApiClassesReference()
#23 {main}
  thrown in /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/BaseActivity.php on line 16
Fatal error: Uncaught Error: Class "Piwik\Plugins\ActivityLog\Activity\Activity" not found in /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/BaseActivity.php:16
Stack trace:
#0 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include()
#1 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#2 /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/VersionBaseActivity.php(12): Composer\Autoload\ClassLoader->loadClass()
#3 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include('...')
#4 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#5 /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/VersionDeleted.php(11): Composer\Autoload\ClassLoader->loadClass()
#6 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(571): include('...')
#7 /var/www/matomo/matomo/developer-documentation/piwik/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile()
#8 [internal function]: Composer\Autoload\ClassLoader->loadClass()
#9 /var/www/matomo/matomo/developer-documentation/generator/ApiClasses/Filter.php(23): class_exists()
#10 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(140): ApiClasses\Filter->acceptClass()
#11 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(100): Sami\Parser\NodeVisitor->addClassOrInterface()
#12 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/NodeVisitor.php(53): Sami\Parser\NodeVisitor->addClass()
#13 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(202): Sami\Parser\NodeVisitor->enterNode()
#14 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(116): PhpParser\NodeTraverser->traverseArray()
#15 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(225): PhpParser\NodeTraverser->traverseNode()
#16 /var/www/matomo/matomo/developer-documentation/generator/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(93): PhpParser\NodeTraverser->traverseArray()
#17 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/CodeParser.php(43): PhpParser\NodeTraverser->traverse()
#18 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Parser/Parser.php(53): Sami\Parser\CodeParser->parse()
#19 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(466): Sami\Parser\Parser->parse()
#20 /var/www/matomo/matomo/developer-documentation/generator/vendor/tsteur/sami/Sami/Project.php(109): Sami\Project->parseVersion()
#21 /var/www/matomo/matomo/developer-documentation/generator/generate.php(119): Sami\Project->update()
#22 /var/www/matomo/matomo/developer-documentation/generator/generate.php(152): generateApiClassesReference()
#23 {main}
  thrown in /var/www/matomo/matomo/developer-documentation/piwik/plugins/TagManager/Activity/BaseActivity.php on line 16
`

Do we expect that the submodules are stable in the parsed branch, or it can happen that sometimes we find an unstable version?
One other thing, not related to this PR, but is there a way to copy and paste a callstack in a GitHub comment and keep the line feeds? Mine are always trimmed and so it's difficult to read them.

**Comment by mattmary** [comment]
Hello @tsteur 

The file which requests the ActivityLog plugin is the `piwik/plugins/AbTesting/Activity/BaseActivity.php` class.

This is class is used a lot:
`piwik/plugins/AbTesting/Activity/ExperimentArchived.php:class ExperimentArchived extends BaseActivity
piwik/plugins/AbTesting/Activity/ExperimentAdded.php:class ExperimentAdded extends BaseActivity
piwik/plugins/AbTesting/Activity/ExperimentDeleted.php:class ExperimentDeleted extends BaseActivity
piwik/plugins/AbTesting/Activity/ExperimentUpdated.php:class ExperimentUpdated extends BaseActivity
piwik/plugins/AbTesting/Activity/BaseActivity.php:abstract class BaseActivity extends Activity
piwik/plugins/AbTesting/Activity/ExperimentFinished.php:class ExperimentFinished extends BaseActivity
piwik/plugins/AbTesting/Activity/ExperimentStarted.php:class ExperimentStarted extends BaseActivity
piwik/plugins/CustomReports/Activity/BaseActivity.php:abstract class BaseActivity extends Activity
piwik/plugins/CustomReports/Activity/ReportAdded.php:class ReportAdded extends BaseActivity
piwik/plugins/CustomReports/Activity/ReportDeleted.php:class ReportDeleted extends BaseActivity
piwik/plugins/CustomReports/Activity/ReportUpdated.php:class ReportUpdated extends BaseActivity
piwik/plugins/FormAnalytics/Activity/FormDeleted.php:class FormDeleted extends BaseActivity
piwik/plugins/FormAnalytics/Activity/FormUpdated.php:class FormUpdated extends BaseActivity
piwik/plugins/FormAnalytics/Activity/BaseActivity.php:abstract class BaseActivity extends Activity
piwik/plugins/FormAnalytics/Activity/FormArchived.php:class FormArchived extends BaseActivity
piwik/plugins/FormAnalytics/Activity/FormAdded.php:class FormAdded extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/HeatmapDeleted.php:class HeatmapDeleted extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/SessionRecordingEnded.php:class SessionRecordingEnded extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/SessionRecordingAdded.php:class SessionRecordingAdded extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/BaseActivity.php:abstract class BaseActivity extends Activity
piwik/plugins/HeatmapSessionRecording/Activity/SessionRecordingDeleted.php:class SessionRecordingDeleted extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/HeatmapEnded.php:class HeatmapEnded extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/HeatmapAdded.php:class HeatmapAdded extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/HeatmapUpdated.php:class HeatmapUpdated extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/SessionRecordingUpdated.php:class SessionRecordingUpdated extends BaseActivity
piwik/plugins/HeatmapSessionRecording/Activity/HeatmapScreenshotDeleted.php:class HeatmapScreenshotDeleted extends BaseActivity
piwik/plugins/TagManager/Activity/VersionDeleted.php:class VersionDeleted extends VersionBaseActivity
piwik/plugins/TagManager/Activity/VersionBaseActivity.php:abstract class VersionBaseActivity extends BaseActivity
piwik/plugins/TagManager/Activity/VariableDeleted.php:class VariableDeleted extends VariableBaseActivity
piwik/plugins/TagManager/Activity/ContainerUpdated.php:class ContainerUpdated extends TagBaseActivity
piwik/plugins/TagManager/Activity/VersionPublished.php:class VersionPublished extends VersionBaseActivity
piwik/plugins/TagManager/Activity/VariableBaseActivity.php:abstract class VariableBaseActivity extends BaseActivity
piwik/plugins/TagManager/Activity/VersionImported.php:class VersionImported extends BaseActivity
piwik/plugins/TagManager/Activity/TriggerBaseActivity.php:abstract class TriggerBaseActivity extends BaseActivity
piwik/plugins/TagManager/Activity/TriggerDeleted.php:class TriggerDeleted extends TriggerBaseActivity
piwik/plugins/TagManager/Activity/TriggerAdded.php:class TriggerAdded extends TriggerBaseActivity
piwik/plugins/TagManager/Activity/VariableUpdated.php:class VariableUpdated extends VariableBaseActivity
piwik/plugins/TagManager/Activity/BaseActivity.php:abstract class BaseActivity extends Activity
piwik/plugins/TagManager/Activity/TagUpdated.php:class TagUpdated extends TagBaseActivity
piwik/plugins/TagManager/Activity/TriggerUpdated.php:class TriggerUpdated extends TriggerBaseActivity
piwik/plugins/TagManager/Activity/TagBaseActivity.php:abstract class TagBaseActivity extends BaseActivity
piwik/plugins/TagManager/Activity/ContainerDeleted.php:class ContainerDeleted extends BaseActivity
piwik/plugins/TagManager/Activity/TagAdded.php:class TagAdded extends TagBaseActivity
piwik/plugins/TagManager/Activity/ContainerAdded.php:class ContainerAdded extends TagBaseActivity
piwik/plugins/TagManager/Activity/VariableAdded.php:class VariableAdded extends VariableBaseActivity
piwik/plugins/TagManager/Activity/TagDeleted.php:class TagDeleted extends TagBaseActivity`

Excluding the Activity pattern from the File search allow generating the documentation.
But I would like your point of view: Does excluding all these files from the documentation generation process is acceptable? Or you want me to go set the filter deeper in the generation process?
Sorry, the plugins structures are quite new for me, so I don't know what is involved.

**Comment by tsteur** [comment]
> But I would like your point of view: Does excluding all these files from the documentation generation process is acceptable? Or you want me to go set the filter deeper in the generation process?

Yes, excluding all these files is acceptable. They aren't needed and we can exclude/filter them.

I'm not fully sure but believe we can exclude directories in https://github.com/matomo-org/developer-documentation/blob/live/generator/generate.php#L57 meaning we could detect all directories that match `plugins/*/Activity` and exclude them.

**Comment by mattmary** [comment]
Hello @tsteur 

Yes, this is what I've done so far to exclude this plugin references.
The documentation has been generated and pushed.

Yes, the CustomAlerts references in the hooks.md have disappeared. But I've checked in the submodule source code, and couldn't find these events. I've also checked that the file is parsed and this is the case.
So I would say this is normal to see disappear these references in the md file. 

But maybe I'm wrong? In this case would you mind highlight me in this plugin source code what is missing?

**Comment by tsteur** [comment]
@mattmary there are no classes showing up of premium features. Do you know why this is happening? These shouldn't be showing up there. AFAIK the premium features should not be on that server where we generate the docs. Also no events should be showing up or any other data of premium features.

image
image


If I see this correctly in hooks.md the CustomAlerts references have not disappeared? It looks to me as that they are showing up.

**Comment by mattmary** [comment]
@tsteur 
I've added the required submodules for the developers when I had the error on the ActivityLog class. Not sure why the git clean -fd did not remove them. Will investigate.
So I've regenerated the documentation. The premium features documentations have disappeared.

I was not expecting that all CustomAlerts references to disappear from the Hooks.md file, just the one which does not exist anymore in the source code. If it is ok for you, it's fine.

**Comment by tsteur** [comment]
Nice @mattmary that looks good 🎉 

note that links from some constants have disappeared like in https://github.com/matomo-org/developer-documentation/pull/633/files#diff-91ec3b5da295707f7064047183d694e67cd1a55def4607565c206ad67b180e6aR330
image

This could be fixed in a follow up pull request though.

**Review by lance-matomo — APPROVED** [review]

**Comment by mattmary** [comment]
@tsteur so is it a go to merge?
Regarding the unresolved links when parsing the source code, do you want me to create an issue in the related projects?

**Comment by tsteur** [comment]
If @lance-matomo has no objections then good to merge 👍  Let me know when we merge so we can plan it as this would be deployed on production the moment we merge this and then also need to adjust the PHP version at the same time in case it doesn't work with PHP 7.4 which is used currently. 
I could also merge this in the next few hours/days and then directly adjust the PHP version if that works for you

**Comment by lance-matomo** [comment]
I'm happy for this to go live, any other problems found can be fixed in follow up tasks. Good effort @mattmary and @tsteur

**Comment by mattmary** [comment]
My question @lance-matomo was if I must create an issue in the related projects for the documentation problems the process raises, or if we ignore them like we did previously.

**Pull request merged**